### PR TITLE
[Snyk] Security upgrade keras from 2.11.0 to 3.13.2

### PR DESCRIPTION
**Pull request by jhamot**
snyk-top-banner

#### Snyk changed the following file(s):

- `tf2/requirements.txt`

<details>
<summary>⚠️ <b>Warning</b></summary>

```
tensorflow 2.11.0 requires keras, which is not installed.
tensorflow 2.11.0 requires tensorflow-io-gcs-filesystem, which is not installed.
tensorflow-metadata 1.12.0 has requirement absl-py<2.0.0,>=0.9, but you have absl-py 2.1.0.
tensorboard 2.11.2 has requirement setuptools>=41.0.0, but you have setuptools 40.5.0.
googleapis-common-protos 1.72.0 has requirement protobuf!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<7.0.0,>=3.20.2, but you have protobuf 3.19.6.
```

</details>

---

> [!IMPORTANT]
>
> - Check the changes in this PR to ensure they won't cause issues with your project.
> - Max score is 1000. Note that the real score may have changed since the PR was raised.
> - This PR was automatically created by Snyk using the credentials of a real user.
> - Some vulnerabilities couldn't be fully fixed and so Snyk will still find them when the project is tested again. This may be because the vulnerability existed within more than one direct dependency, but not all of the affected dependencies could be upgraded.

---

**Note:** _You are seeing this because you or someone else with access to this repository has authorized Snyk to open fix PRs._

For more information: 
🧐 [View latest project report](https://app.snyk.io/org/lucasfilm/project/c1bd5c26-cdb9-4fef-9b86-a38cffefa71b?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr)
📜 [Customise PR templates](https://docs.snyk.io/scan-using-snyk/pull-requests/snyk-fix-pull-or-merge-requests/customize-pr-templates?utm_source=github&utm_content=fix-pr-template)
🛠 [Adjust project settings](https://app.snyk.io/org/lucasfilm/project/c1bd5c26-cdb9-4fef-9b86-a38cffefa71b?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr/settings)
📚 [Read about Snyk's upgrade logic](https://docs.snyk.io/scan-with-snyk/snyk-open-source/manage-vulnerabilities/upgrade-package-versions-to-fix-vulnerabilities?utm_source=github&utm_content=fix-pr-template)

---

**Learn how to fix vulnerabilities with free interactive lessons:**

🦉 [Learn about vulnerability in an interactive lesson of Snyk Learn.](https://learn.snyk.io/?loc&#x3D;fix-pr)

[//]: # 'snyk:metadata:{"breakingChangeRiskLevel":null,"FF_showPullRequestBreakingChanges":false,"FF_showPullRequestBreakingChangesWebSearch":false,"customTemplate":{"variablesUsed":[],"fieldsUsed":[]},"dependencies":[{"name":"keras","from":"2.11.0","to":"3.13.2"}],"env":"prod","issuesToFix":["SNYK-PYTHON-KERAS-15268069"],"prId":"b0696305-7b2e-4f7c-9849-9f1653fe3eb8","prPublicId":"b0696305-7b2e-4f7c-9849-9f1653fe3eb8","packageManager":"pip","priorityScoreList":[116],"projectPublicId":"c1bd5c26-cdb9-4fef-9b86-a38cffefa71b","projectUrl":"https://app.snyk.io/org/lucasfilm/project/c1bd5c26-cdb9-4fef-9b86-a38cffefa71b?utm_source=github&utm_medium=referral&page=fix-pr","prType":"fix","templateFieldSources":{"branchName":"default","commitMessage":"default","description":"default","title":"default"},"templateVariants":["updated-fix-title","pr-warning-shown","priorityScore"],"type":"auto","upgrade":[],"vulns":["SNYK-PYTHON-KERAS-15268069"],"patch":[],"isBreakingChange":false,"remediationStrategy":"vuln"}'

### CODE REVIEW

**Comment by jhamot** [comment]
Merge Risk: High

This is a major version upgrade from Keras 2 to Keras 3, which represents a fundamental architectural shift. Keras has been rewritten as a multi-backend framework that can run on top of TensorFlow, JAX, and PyTorch, and is no longer just a high-level API for TensorFlow.  
  
**This is a high-risk upgrade that requires significant code and environment changes.**  
  
### Key Breaking Changes:  
  
*   **Multi-Backend Architecture**: Keras is now a standalone package. The tight integration with TensorFlow (`tf.keras`) is gone. To continue using the TensorFlow backend, you must configure it before importing Keras.  
  
*   **Python Version Requirement**: Keras 3.13.0 and newer require **Python 3.11 or higher**.  
  
*   **Mandatory Import Changes**: All import statements must be updated.  
    *   `from tensorflow.keras import layers` becomes `from keras import layers`.  
    *   `import tensorflow.keras` becomes `import keras`.  
  
*   **Optimizer API Changes**: The optimizers from Keras 2 have been moved to the `tf.keras.optimizers.legacy` namespace. Using the new `keras.optimizers` may break checkpoint compatibility.  
  
*   **Default Save Format**: The default format for `model.save()` is now the native Keras v3 (`.keras`) format, not HDF5 (`.h5`).  
  
*   **API Refactoring for Portability**: While Keras 3 aims for backward compatibility when using the TensorFlow backend, any code using TensorFlow-specific APIs (e.g., `tf.nn`, `tf.linalg`) must be refactored to use the new `keras.ops` namespace to be truly backend-agnostic.  
  
**Recommendation:**  
This is a major migration effort. Developers must:  
1.  Ensure their environment uses Python 3.11+.  
2.  Update all Keras import statements throughout the codebase.  
3.  Review and update optimizer usage, especially if loading old checkpoints.  
4.  Verify model saving and loading logic to account for the new default format.  
5.  Consider refactoring TensorFlow-specific code to use `keras.ops` to leverage the benefits of the new multi-backend architecture.  
  
For projects not ready for this migration, an alternative is to use the `tf_keras` package, which is a maintained version of Keras 2.  
  
Source: [Keras 2 to 3 Migration Guide](https://keras.io/guides/migrating_to_keras_3/), [Keras 3 Release Announcement](https://keras.io/keras_3/)

> **Notice 🤖:** This content was augmented using **artificial intelligence**. AI-generated content **may** contain **errors** and should be **reviewed** for accuracy before use.
